File: ios/AtlasCore/Sources/AtlasCore/Keys/Recovery.swift
import Foundation
import CommonCrypto

/// Threshold recovery — STRATIFIED by release mechanism (§7.2, §7.3). Mirrors
/// `backend/atlas/keys/recovery.py`.
///
///   * DEVICE-PRESENT (card, in-person, normal auth) -> Secure Enclave robust
///     biometric release of `share_bio` (device-bound).
///   * TOTAL-LOSS / catastrophic -> the two PORTABLE shares (card + context),
///     NO Enclave and NO biometric. Fuzzy extractor RETIRED (TRUST_LAYER.md #7):
///     the anti-spoof at total loss is the in-person recovery person.
///
/// Invariants: never store the biometric (Enclave-sealed only); 2-of-3 threshold
/// unchanged; total-loss never depends on a single device's Enclave (no Enclave param).
public enum RecoveryError: Error { case attestationRequired, enclaveReleaseFailed,
                                        noTrustedContext, unknownHandle, badPasscode(remaining: Int),
                                        attemptsExhausted,
                                        /// Holder-authority gate (§6): recovery is triggered ONLY by the
                                        /// user's own authority — no operator, court, or system path. Same
                                        /// discipline as rerooting's OperatorForbidden.
                                        holderAuthorityRequired }

/// Passcode KDF: salted + stretched. The recovery-child passcode is a low-entropy
/// selector (the real secrecy is biometric + threshold), but it must still resist
/// offline brute force if the enrolment record leaks — so never store a bare hash.
private let passcodeIters: UInt32 = 100_000

/// PBKDF2-HMAC-SHA256 (matches Python `hashlib.pbkdf2_hmac("sha256", ..., 100_000)`,
/// default 32-byte output). CryptoKit has no PBKDF2, so use CommonCrypto.
private func derivePasscode(_ passcode: String, salt: Data) -> Data {
    let pwd = Data(passcode.utf8)
    var derived = Data(count: 32)
    let status = derived.withUnsafeMutableBytes { (dk: UnsafeMutableRawBufferPointer) -> Int32 in
        salt.withUnsafeBytes { (saltPtr: UnsafeRawBufferPointer) -> Int32 in
            pwd.withUnsafeBytes { (pwdPtr: UnsafeRawBufferPointer) -> Int32 in
                CCKeyDerivationPBKDF(
                    CCPBKDFAlgorithm(kCCPBKDF2),
                    pwdPtr.bindMemory(to: Int8.self).baseAddress, pwd.count,
                    saltPtr.bindMemory(to: UInt8.self).baseAddress, salt.count,
                    CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                    passcodeIters,
                    dk.bindMemory(to: UInt8.self).baseAddress, 32)
            }
        }
    }
    precondition(status == kCCSuccess, "PBKDF2 derivation failed")
    return derived
}

/// Constant-time byte compare (mirrors Python `hmac.compare_digest`).
private func constantTimeEquals(_ a: Data, _ b: Data) -> Bool {
    guard a.count == b.count else { return false }
    var diff: UInt8 = 0
    for (x, y) in zip(a, b) { diff |= x ^ y }
    return diff == 0
}

/// Artifacts produced once, at enrolment (§7.2). Reference type: the lockout
/// counter is PERSISTED HERE (not on the gate object) so re-instantiating the
/// gate does NOT reset the 3-attempt limit.
public final class RecoveryEnrolment {
    public let shareCard: Shamir.Share          // portable; JavaCard
    public let shareContext: Shamir.Share       // portable; trusted-context vertex
    public let enclaveDeviceID: Data            // device-present release (Enclave)
    public let enclaveSealedBio: Data
    public let recoveryChildHandle: Data
    let passcodeSalt: Data
    let passcodeHash: Data
    // Lockout counter lives HERE (the persisted record), not on the gate object —
    // otherwise an attacker resets the limit by re-instantiating the gate.
    var childAttemptsRemaining: Int = 3

    init(shareCard: Shamir.Share, shareContext: Shamir.Share, enclaveDeviceID: Data,
         enclaveSealedBio: Data, recoveryChildHandle: Data, passcodeSalt: Data, passcodeHash: Data) {
        self.shareCard = shareCard; self.shareContext = shareContext
        self.enclaveDeviceID = enclaveDeviceID; self.enclaveSealedBio = enclaveSealedBio
        self.recoveryChildHandle = recoveryChildHandle
        self.passcodeSalt = passcodeSalt; self.passcodeHash = passcodeHash
    }
}

public enum Recovery {
    static let bioLabel = Data("share-bio".utf8)

    public static func enrol(_ tree: IdentityTree, biometricTemplate: Data,
                             device: BiometricEnclave, passcode: String,
                             sphincs: SphincsProvider) throws -> RecoveryEnrolment {
        let shares = Shamir.split(tree.tskSeed, n: 3, k: 2)
        let shareBio = shares[1]
        // Device-present: enrol biometric + seal share_bio to this device.
        device.enrolBiometric(biometricTemplate)
        let enclaveSealed = device.seal(shareBio.encode(), label: bioLabel)
        let salt = Primitives.randomBytes(16)
        return RecoveryEnrolment(
            shareCard: shares[0], shareContext: shares[2],
            enclaveDeviceID: device.deviceID, enclaveSealedBio: enclaveSealed,
            recoveryChildHandle: tree.child(.recovery).handle,
            passcodeSalt: salt,
            passcodeHash: derivePasscode(passcode, salt: salt))
    }

    // MARK: Device-present — Secure Enclave robust biometric release

    private static func enclaveBioShare(_ enr: RecoveryEnrolment, _ device: BiometricEnclave,
                                        _ liveSample: Data) -> Shamir.Share? {
        guard device.deviceID == enr.enclaveDeviceID else { return nil }  // device-bound
        guard let raw = device.release(enr.enclaveSealedBio, liveSample: liveSample, label: bioLabel)
        else { return nil }
        return Shamir.Share.decode(raw)
    }

    public static func recoverViaCard(_ enr: RecoveryEnrolment, device: BiometricEnclave,
                                      cardShare: Shamir.Share, liveBiometric: Data, attested: Bool,
                                      userAuthorized: Bool, sphincs: SphincsProvider) throws -> IdentityTree {
        guard userAuthorized else { throw RecoveryError.holderAuthorityRequired }
        guard attested else { throw RecoveryError.attestationRequired }
        guard let bio = enclaveBioShare(enr, device, liveBiometric) else { throw RecoveryError.enclaveReleaseFailed }
        return try IdentityTree.build(tskSeed: Shamir.combine([cardShare, bio]), sphincs: sphincs)
    }

    public static func recoverInPerson(_ enr: RecoveryEnrolment, device: BiometricEnclave,
                                       liveBiometric: Data, attested: Bool, inPersonTrustedContext: Bool,
                                       userAuthorized: Bool, sphincs: SphincsProvider) throws -> IdentityTree {
        guard userAuthorized else { throw RecoveryError.holderAuthorityRequired }
        guard attested else { throw RecoveryError.attestationRequired }
        guard inPersonTrustedContext else { throw RecoveryError.noTrustedContext }
        guard let bio = enclaveBioShare(enr, device, liveBiometric) else { throw RecoveryError.enclaveReleaseFailed }
        return try IdentityTree.build(tskSeed: Shamir.combine([bio, enr.shareContext]), sphincs: sphincs)
    }

    /// Normal auth: robust device-present human-proof (no secret leaves).
    public static func releaseForAuth(_ enr: RecoveryEnrolment, device: BiometricEnclave,
                                      liveBiometric: Data) -> Bool {
        enclaveBioShare(enr, device, liveBiometric) != nil
    }

    // MARK: Total-loss — the two PORTABLE threshold shares (NO Enclave, NO biometric)

    /// Recovery from TOTAL DEVICE LOSS on a NEW device: combine the card share (Half B, which
    /// you carry) and the trusted-context vertex (released only under the in-person recovery
    /// ceremony) into 2-of-3. NO Enclave and NO biometric — the anti-spoof is the live,
    /// accountable recovery person of the in-person ceremony (see recovery_anchor).
    public static func recoverTotalLoss(_ enr: RecoveryEnrolment, cardShare: Shamir.Share,
                                        contextShare: Shamir.Share, attested: Bool,
                                        inPersonTrustedContext: Bool,
                                        userAuthorized: Bool, sphincs: SphincsProvider) throws -> IdentityTree {
        guard userAuthorized else { throw RecoveryError.holderAuthorityRequired }
        guard attested else { throw RecoveryError.attestationRequired }
        guard inPersonTrustedContext else { throw RecoveryError.noTrustedContext }
        return try IdentityTree.build(tskSeed: Shamir.combine([cardShare, contextShare]),
                                      sphincs: sphincs)
    }
}

/// Private handle + 3-attempt passcode -> recovery child capability only (§7.3).
///
/// The attempt counter is PERSISTED in the enrolment record (`enr`), not on this
/// object — so re-instantiating the gate does NOT reset the lockout (that bypass
/// would otherwise turn the 3-attempt limit into unlimited guesses). A correct
/// passcode does not consume an attempt; only failures do.
public final class RecoveryChildGate {
    private let enr: RecoveryEnrolment
    public init(_ enr: RecoveryEnrolment) { self.enr = enr }

    public var attemptsRemaining: Int { enr.childAttemptsRemaining }

    public func attempt(assertedHandle: Data, passcode: String, attested: Bool,
                        userAuthorized: Bool = true) throws -> Data {
        guard userAuthorized else { throw RecoveryError.holderAuthorityRequired }
        guard attested else { throw RecoveryError.attestationRequired }
        guard enr.childAttemptsRemaining > 0 else { throw RecoveryError.attemptsExhausted }
        guard assertedHandle == enr.recoveryChildHandle else {
            enr.childAttemptsRemaining -= 1
            throw RecoveryError.unknownHandle
        }
        let candidate = derivePasscode(passcode, salt: enr.passcodeSalt)
        guard constantTimeEquals(candidate, enr.passcodeHash) else {
            enr.childAttemptsRemaining -= 1
            throw RecoveryError.badPasscode(remaining: enr.childAttemptsRemaining)
        }
        return enr.recoveryChildHandle
    }
}

File: backend/atlas/keys/recovery.py
"""Threshold recovery — STRATIFIED by release mechanism (§7.2, §7.3).

The TSK seed is split 2-of-3 (Shamir) into three vertices: `share_card`,
`share_bio`, `share_context`.

  * DEVICE-PRESENT paths (card, in-person, normal auth) release `share_bio` via
    the Secure Enclave (robust biometric match, works on real fingers/faces).
    The Enclave-sealed copy is DEVICE-BOUND.
  * TOTAL-LOSS / catastrophic path (new device; the old device is gone) recovers
    from the two PORTABLE shares — `share_card` (JavaCard / Half B, which you carry)
    and `share_context` (the trusted-context / backend vertex, released only under
    the in-person recovery ceremony). It takes NO Enclave and NO biometric.

Fuzzy extractor RETIRED (2026-07-17, TRUST_LAYER.md #7): Atlas extracts no key from
raw biometrics and stores no biometric sketch. Biometric matching is the Secure
Enclave (device-present) or a LIVE recovery person (total loss — see
`realid.recovery_anchor`); at total loss the anti-spoof is that accountable human,
not a reconstructed key.

Why stratify: the Enclave is robust but device-bound, so it CANNOT be the total-loss
path (losing the device loses the sealed key). Total loss therefore rides the two
portable threshold shares, which by construction do not depend on the lost device.

Invariants preserved:
  * "Never store the biometric" — the Enclave keeps the template sealed and matches
    internally; nothing biometric is persisted outside it.
  * 2-of-3 threshold — unchanged.
  * Total-loss recovery NEVER depends on a single device's Enclave (takes no Enclave).

Precondition for every path (§7.3): attested HW+FW + (device-present) live biometric.
"""

from __future__ import annotations

import hashlib
import hmac
from dataclasses import dataclass, field

from ..crypto import shamir
from ..crypto.primitives import random_bytes
from .enclave import SecureEnclave
from .identity import IdentityTree, build_identity_tree

# Passcode KDF: salted + stretched. The recovery-child passcode is a low-entropy
# selector (the real secrecy is biometric + threshold), but it must still resist
# offline brute force if the enrolment record leaks — so never store a bare hash.
_PASSCODE_ITERS = 100_000


def _derive_passcode(passcode: str, salt: bytes) -> bytes:
    return hashlib.pbkdf2_hmac("sha256", passcode.encode(), salt, _PASSCODE_ITERS)


class RecoveryError(RuntimeError):
    pass


class AttestationRequired(RecoveryError):
    """Precondition (§7.3): every path requires attested HW+FW."""


class HolderAuthorityRequired(RecoveryError):
    """Holder-authority gate (Credential PQC Posture §6): recovery is triggered
    ONLY by the user's own authority (unlock / in-person ceremony). There is no
    operator, court, or system path — same discipline as `rerooting.py`'s
    OperatorForbidden. Modelled here as an explicit `user_authorized` flag that
    no operator code path can set; the real binding is the live biometric +
    threshold + in-person ceremony the paths already require."""


_BIO_LABEL = b"share-bio"


@dataclass
class RecoveryEnrolment:
    """Artifacts produced once, at enrolment (§7.2)."""

    share_card: shamir.Share          # portable; goes on the JavaCard
    share_context: shamir.Share       # portable; trusted-context / backend vertex
    # device-present release of share_bio (Secure Enclave, robust, device-bound):
    enclave_device_id: bytes
    enclave_sealed_bio: bytes
    recovery_child_handle: bytes
    _passcode_salt: bytes = field(repr=False, default=b"")
    _passcode_hash: bytes = field(repr=False, default=b"")
    # Lockout counter lives HERE (the persisted record), not on the gate object —
    # otherwise an attacker resets the limit by re-instantiating the gate.
    _child_attempts_remaining: int = 3


def enrol_recovery(
    tree: IdentityTree, biometric_template: bytes, *, device: SecureEnclave, passcode: str
) -> RecoveryEnrolment:
    """Split the TSK seed 2-of-3 and bind `share_bio` to the device Enclave
    (device-present). Total-loss recovery uses the two PORTABLE shares (card +
    context) and needs no biometric release of `share_bio`."""
    share_card, share_bio, share_context = shamir.split(tree.tsk_seed, n=3, k=2)

    # Device-present: enrol the biometric in the Enclave and seal share_bio to it.
    device.enrol_biometric(biometric_template)
    enclave_sealed_bio = device.seal(share_bio.encode(), label=_BIO_LABEL)

    return RecoveryEnrolment(
        share_card=share_card,
        share_context=share_context,
        enclave_device_id=device.device_id,
        enclave_sealed_bio=enclave_sealed_bio,
        recovery_child_handle=tree.child("recovery").handle,
        _passcode_salt=(salt := random_bytes(16)),
        _passcode_hash=_derive_passcode(passcode, salt),
    )


def _rebuild(seed: bytes) -> IdentityTree:
    return build_identity_tree(seed)


# ---------------------------------------------------------------------------
# DEVICE-PRESENT paths — Secure Enclave biometric release (robust, device-bound)
# ---------------------------------------------------------------------------

def _enclave_bio_share(enr: RecoveryEnrolment, device: SecureEnclave, live_sample: bytes):
    if device.device_id != enr.enclave_device_id:
        # device-bound: only the enrolled device's Enclave can release it
        return None
    raw = device.release(enr.enclave_sealed_bio, live_sample=live_sample, label=_BIO_LABEL)
    return shamir.Share.decode(raw) if raw is not None else None


def recover_via_card(
    enr: RecoveryEnrolment, *, device: SecureEnclave, card_share: shamir.Share,
    live_biometric: bytes, attested: bool, user_authorized: bool,
) -> IdentityTree:
    """Remote self-service on the ENROLLED device: card share + Enclave-released
    biometric share -> 2-of-3 (§7.3). Robust matching (works on real fingers)."""
    if not user_authorized:
        raise HolderAuthorityRequired("card path requires the user's own authority (no operator path)")
    if not attested:
        raise AttestationRequired("card path requires attested HW+FW")
    bio_share = _enclave_bio_share(enr, device, live_biometric)
    if bio_share is None:
        raise RecoveryError("Enclave biometric release failed (no match / wrong device)")
    return _rebuild(shamir.combine([card_share, bio_share]))


def recover_in_person(
    enr: RecoveryEnrolment, *, device: SecureEnclave, live_biometric: bytes,
    attested: bool, in_person_trusted_context: bool, user_authorized: bool,
) -> IdentityTree:
    """In-person forward-recovery on the ENROLLED device: Enclave-released
    biometric share + trusted-context vertex -> 2-of-3 (§7.3). The blind
    System-ID forward-recovers everything; the root stays blind."""
    if not user_authorized:
        raise HolderAuthorityRequired("in-person path requires the user's own authority (no operator path)")
    if not attested:
        raise AttestationRequired("in-person path requires attested HW+FW")
    if not in_person_trusted_context:
        raise RecoveryError("in-person trusted context not established")
    bio_share = _enclave_bio_share(enr, device, live_biometric)
    if bio_share is None:
        raise RecoveryError("Enclave biometric release failed (no match / wrong device)")
    return _rebuild(shamir.combine([bio_share, enr.share_context]))


def release_for_auth(enr: RecoveryEnrolment, *, device: SecureEnclave, live_biometric: bytes) -> bool:
    """Normal auth: the Enclave robustly releases on a live biometric match.
    Returns whether the device-present human-proof succeeded (no secret leaves)."""
    return _enclave_bio_share(enr, device, live_biometric) is not None


# ---------------------------------------------------------------------------
# TOTAL-LOSS / catastrophic path — the two PORTABLE threshold shares.
# Takes NO Enclave and NO biometric: by construction it cannot depend on the
# lost device, and the anti-spoof is the in-person recovery ceremony.
# ---------------------------------------------------------------------------

def recover_total_loss(
    enr: RecoveryEnrolment, *, card_share: shamir.Share, context_share: shamir.Share,
    attested: bool, in_person_trusted_context: bool, user_authorized: bool,
) -> IdentityTree:
    """Recovery from TOTAL DEVICE LOSS on a NEW device (§7.3, catastrophic).

    Combines the two PORTABLE threshold shares — `card_share` (Half B, which you
    carry) and `context_share` (the trusted-context / backend vertex, released only
    under the in-person recovery ceremony) — into 2-of-3. NEVER touches an Enclave
    and uses NO biometric: identity assurance at total loss is the live, accountable
    recovery person of the in-person ceremony (see `realid.recovery_anchor`), not a
    stored template or a reconstructed key.
    """
    if not user_authorized:
        raise HolderAuthorityRequired("total-loss recovery requires the user's own authority (no operator path)")
    if not attested:
        raise AttestationRequired("total-loss path requires attested HW+FW (new device)")
    if not in_person_trusted_context:
        raise RecoveryError("total-loss recovery is in-person only")
    return _rebuild(shamir.combine([card_share, context_share]))


# ---------------------------------------------------------------------------
# Recovery-child entry — private handle + 3-attempt passcode (selector/gate).
# ---------------------------------------------------------------------------

@dataclass
class RecoveryChildSession:
    handle: bytes
    attempts_remaining: int


class RecoveryChildGate:
    """Private handle + 3-attempt passcode (selector/gate, not the security; §7.3).

    The attempt counter is PERSISTED in the enrolment record (`enr`), not on this
    object — so re-instantiating the gate does NOT reset the lockout (that bypass
    would otherwise turn the 3-attempt limit into unlimited guesses). A correct
    passcode does not consume an attempt; only failures do.
    """

    def __init__(self, enr: RecoveryEnrolment):
        self._enr = enr

    @property
    def attempts_remaining(self) -> int:
        return self._enr._child_attempts_remaining

    def attempt(self, *, asserted_handle: bytes, passcode: str, attested: bool,
                user_authorized: bool = True) -> RecoveryChildSession:
        if not user_authorized:
            raise HolderAuthorityRequired("recovery-child path requires the user's own authority")
        if not attested:
            raise AttestationRequired("recovery-child path requires attested HW+FW")
        if self._enr._child_attempts_remaining <= 0:
            raise RecoveryError("passcode attempts exhausted; full recovery required")
        if asserted_handle != self._enr.recovery_child_handle:
            self._enr._child_attempts_remaining -= 1
            raise RecoveryError("unknown recovery-child handle")
        candidate = _derive_passcode(passcode, self._enr._passcode_salt)
        if not hmac.compare_digest(candidate, self._enr._passcode_hash):
            self._enr._child_attempts_remaining -= 1
            raise RecoveryError(f"bad passcode ({self._enr._child_attempts_remaining} attempts left)")
        return RecoveryChildSession(
            handle=self._enr.recovery_child_handle,
            attempts_remaining=self._enr._child_attempts_remaining,
        )
